### ISSUE TITLE
lossgradient returns nothing

### ISSUE THREAD

**Issue by jeffreylu9** [denizyuret/Knet.jl#221]
I wrote a custom loss function that returns a Float32, and I wrote code below that makes use of the grad function:

```
loss(w,x,ygold) = crb_loss(predict(w,x), ygold)
lossgradient = grad(loss)
```

However, when using lossgradient to update weights as shown below, the lossgradient function returns nothing. Any reason why this might be happening?

```
function train(w, data; lr=.1, epochs=3, iters=1800)
    for epoch=1:epochs
        for (x,y) in data
            g = lossgradient(w, x, y)
            update!(w, g, lr=lr)
            if (iters -= 1) <= 0
                return w
            end
        end
    end
    return w
end
```

**Comment by denizyuret** [comment]
This may happen when your weights have no effect on the loss.

**Comment by jeffreylu9** [comment]
I think our weights do have an effect though, since after random initialization of the weights the loss changes.

**Comment by denizyuret** [comment]
More precisely small changes in weights may not have an effect on the
output.  Gradient only measures changes in the local neighborhood, so if
the function is flat in the local neighborhood you will get zero gradients
which possibly explains the situation.  If your output changes in response
to epsilon changes in weights then I need to look into it further.

**Comment by jeffreylu9** [comment]
Moved discussion to https://groups.google.com/forum/#!topic/knet-users/RrGgjSiHuJU
